L74-java-channel-creds.md

```diff
@@ -0,0 +1,489 @@
+   * credentials if no configuration is provided.
+   *
+   * @throws IllegalArgumentException if fallback is unable to be used
+   */
+  public static ChannelCredentials createWithFallback(
+      ChannelCredentials fallback) {
+    return new InternalNettyChannelCredentials.create(
+        new XdsProtocolNegotiatorFactory(
+          InternalNettyChannelCredentials.toNegotiator(fallback)));
+  }
+}
+```
+
+### Netty ChannelCredentials Processing Example
+
+The API provided here is not part of the design. It is meant as a
+demonstration.
+
+```java
+final class ProtocolNegotiators {
+  private static final EnumSet<> understoodTlsFeatures =
+      EnumSet.noneOf(TlsChannelCredentials.Feature.class);
+
+  public static Result from(ChannelCredentials creds) {
+    if (creds instanceof TlsChannelCredentials) {
+      TlsChannelCredentials tlsCreds = (TlsChannelCredentials) creds;
+      Set<TlsChannelCredentials.Feature> incomprehensible =
+          tlsCreds.incomprehensible(understoodTlsFeatures);
+      if (!incomprehensible.isEmpty()) {
+        return Result.error("TLS features not understood: " + incomprehensible);
+      }
+      return Result.negotiator(tls());
+
+    } else if (creds instanceof InsecureChannelCredentials) {
+      return Result.negotiator(plaintext());
+
+    } else if (creds instanceof CompositeChannelCredentials) {
+      CompositeChannelCredentials compCreds = (CompositeChannelCredentials) creds;
+      return from(compCreds.getChannelCredentials())
+          .withCallCredentials(compCreds.getCallCredentials());
+
+    } else if (creds instanceof NettyChannelCredentials) {
+      NettyChannelCredentials nettyCreds = (NettyChannelCredentials) creds;
+      return Result.negotiator(nettyCreds.getNegotiator());
+
+    } else if (creds instanceof ChoiceChannelCredentials) {
+      ChoiceChannelCredentials choiceCreds = (ChoiceChannelCredentials) creds;
+      StringBuilder error = new StringBuilder();
+      for (ChannelCredentials innerCreds : choiceCreds.getCredentialsList()) {
+        Result result = from(innerCreds);
+        if (!result.isError()) {
+          return result;
+        }
+        error.append("; ");
+        error.append(result.getError());
+      }
+      return Result.error(error.substring(2));
+
+    } else {
+      return Result.error(
+          "Unsupported credential type: " + creds.getClass().getName());
+    }
+  }
+
+  public final class Result {
+    public static Result error(String error) {...}
+    public static Result negotiator(ProtocolNegotiator.Factory factory) {...}
+    public Result withCallCredentials(CallCredentials callCreds) {...}
+  }
+}
+```
+
+### ServerCredentials
+
+Server-side will be a clone of client-side, but using the `Server` instead of
+`Channel`. `ServerCredentials` cannot contain `CallCredentials`, so
+`CompositeServerCredentials` will be omitted. There is also currently no known
+use-case for `ChoiceServerCredentials`, so it will be omitted.
```

Comment:
I am wondering if `ChoiceServerCredentials` can be the API for the SNI use-case. So each ChannelCredentials in the list will be some kind of TlsServerCredentials with the associated hostname (`javax.net.ssl.SNIServerName` ?) and the implementation will pick the correct TlsServerCredentials for the associated hostName in the incoming connection. An alternative is to have a special feature in TlsServerCredentials that takes a map of SNIServerName to key/cert pair but looks more complex (and less flexible?) to me.

Reply:
Probably not. I think we should stick with the pattern where the credentials are converted into some internal representation at channel creation time. I'm not excited about the conversion occurring in the data path for performance and complexity reasons. Also, SNI we'd have serious issues supporting SNI if it could change arbitrary TLS configuration. I know xDS has some stuff along those lines, but I think we'd handle anything like that with a low-level API (protocol negotiator) and not at a high level here.

Reply:
yes, I do understand where you are coming from and I sort of expected that. I remember there was some kind of comment exchange with @ZhenLian where he said gRPC core/C++ TLS credentials API has a way (or may be a proposal) of providing a map of hostname->certs to support SNI so I thought we could match that.But low level API sounds good for now